xbrowse/annotation/vep_annotations.py

```diff
@@ -257,47 +257,64 @@ def get_csq_fields_from_vcf_desc(csq_desc):
-    if num_annotations == 0:
-        print 'Warning: no VEP annnotation'
-        return None
+    if not transcript_annotations:
+        raise ValueError("transcript_annotations is empty")
 
-    worst_value = 1000
-    worst_index = -1
-    for i in range(num_annotations):
+    annotations = [(i, ta) for i, ta in enumerate(transcript_annotations)]
 
-        if gene_id and vep_annotation[i]['gene'] != gene_id: continue
-
-        annot = vep_annotation[i]['consequence']
+    # filter by gene_id
+    if gene_id is not None:
+        annotations = [(i, ta) for i, ta in annotations if ta['gene'] == gene_id]
+        if not annotations:
+            raise ValueError("None of the transcripts in %s have gene_id: %s" % (transcript_annotations, gene_id))
 
+    # find the transcript(s) affected with the worst severity
+    worst_severity = 10**9   # lower numbers are worse severity
+    worst_severity_annotations = []  # a list of worst-severity transcripts
+    for i, transcript_annotation in annotations:
         try:
-            pos = SO_SEVERITY_ORDER.index(annot)
-
-            # hack: this is to deprioritize noncoding and nonsense mediated decay transcripts
-            if vep_annotation[i]['is_nc']:
-                pos += NUM_SO_TERMS
-            if vep_annotation[i]['is_nmd']:
-                pos += 2*NUM_SO_TERMS
-
-        except ValueError:
-            print 'Warning: no VEP ordering for %s' % annot
-            return None
-
-        if pos < worst_value:
-            worst_index = i
-            worst_value = pos
-
-    return worst_index
+            severity_scale = SO_SEVERITY_ORDER.index(
+                transcript_annotation['consequence'])
+        except ValueError as e:
+            raise ValueError("Unexpected VEP consequence: %s: %s" % (
+                transcript_annotation['consequence'], e))
+
+        # hack: this is to deprioritize noncoding and nonsense mediated decay transcripts
+        if transcript_annotation['is_nc']:
+            severity_scale += NUM_SO_TERMS
+        if transcript_annotation['is_nmd']:
+            severity_scale += 2*NUM_SO_TERMS
+
+        if severity_scale <= worst_severity:
+            if severity_scale < worst_severity:
+                worst_severity_annotations = []
+                worst_severity = severity_scale
+
+            worst_severity_annotations.append((i, transcript_annotation))
+
+    # if multiple transcripts have the same worst severity, chose the canonical transcript
+    for i, transcript_annotation in worst_severity_annotations:
+        if transcript_annotation['canonical']:
+            worst_severity_annotation_index = i
+            break
+    else:
+        # otherwise, sort transcripts alphabetically by transcript id and return the 1st one
+        worst_severity_annotations.sort(key=lambda x: x[1]['feature'])
+        worst_severity_annotation_index, _ = worst_severity_annotations[0]
+
+    return worst_severity_annotation_index
```

Comment:
As a small stylistic comment, I generally stick to small single task functions vs longer functions to help me partition tasks so it is easier to debug, but the trade-off of is that it causes having a lot of smaller functions floating around.